### PR TITLE
Introduce the asList method in ConfigOptions

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

Introduce the asList method in ConfigOptions

### Why are the changes needed?
It’s necessary when getting the list of values from the config directly.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by zuston** [comment]
@jerqi Could u help review?

**Comment by jerqi** [comment]
Our RssConf imitate SparkConf's style. SparkConf don't have this function, too. Is it better to split config options on the outside of RssConf?

**Comment by zuston** [comment]
I think it looks RssConf imitate FlinkConf's style.

File: common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java [hunk 1]
```diff
@@ -208,4 +217,69 @@ public ConfigOption<T> noDefaultValue() {
...
+      this.asListConverter = (v) -> {
+        return Arrays.stream(v.toString().split(LIST_SPILTTER))
+                .map(s -> atomicConverter.apply(s)).collect(Collectors.toList());
+      };
+    }
+
+    public ListConfigOptionBuilder checkValue(Function<E, Boolean> checkValueFunc, String errMsg) {
+      Function<Object, List<E>> newConverter = (v) -> {
+        List<E> list = Arrays.stream(v.toString().split(LIST_SPILTTER))
+                .map(s -> atomicConverter.apply(s)).collect(Collectors.toList());
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java:245]: Just curious, Do Flink support the method `checkValue`?

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java [hunk 2]
```diff
@@ -581,6 +582,11 @@ public <T> RssConf set(ConfigOption<T> option, T value) {
     return this;
   }
 
+  public <T> RssConf setList(ConfigOption<List<T>> option, Object value) {
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/RssConf.java:585]: Flink class `Configuration` don't have method `setList`. Could we imitate the Flink api?

**Comment by jerqi** [comment]
> I think it looks RssConf imitate FlinkConf's style. In Flink, it has asList method.

OK.

File: common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java (hunk 1 above)
Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java:245]: Flink don’t support checkvalue. But it is added in uniffle TypedConfigOptionBuilder. And i think it’s useful and add it into ListConfigOptionBuilder

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java (hunk 2 above)
Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java:585]: Let me take a look.

Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]: You are right. I remove it and reuse the method of `RssConf.set` to set list values.

**Comment by zuston** [comment]
Updated. @jerqi

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java [hunk 3]
```diff
@@ -576,7 +576,7 @@ public <T> Optional<T> getOptional(ConfigOption<T> option) {
     return value;
   }
 
-  public <T> RssConf set(ConfigOption<T> option, T value) {
+  public <T> RssConf set(ConfigOption<T> option, Object value) {
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/RssConf.java:579]: Why do we use Object? It's better to use class `T`.

File: common/src/test/java/com/tencent/rss/common/config/ConfigOptionTest.java [hunk 4]
```diff
@@ -22,12 +22,88 @@
...
+            .asList()
+            .defaultValues("h1", "h2")
+            .withDescription("List config key1");
+
+    List<String> defaultValues = listStringConfigOption.defaultValue();
+    assertEquals(2, defaultValues.size());
+    assertSame(String.class, listStringConfigOption.getClazz());
+
+    RssBaseConf conf = new RssBaseConf();
+    conf.set(listStringConfigOption, "a,b,c");
```
[the remark this answers is not in the document]
Comment (jerqi) [reply common/src/test/java/com/tencent/rss/common/config/ConfigOptionTest.java:50]: And we should add a case like :
```
conf.set(listStringOption, List("a", "b", c))
assertEquals("[a , b, c]", conf.get(listStringOption).toString());
```

File: common/src/test/java/com/tencent/rss/common/config/ConfigOptionTest.java [hunk 5]
```diff
@@ -22,12 +22,88 @@
...
+            .asList()
+            .defaultValues(1)
+            .withDescription("List long config key2");
+
+    List<Long> longDefaultVals = listLongConfigOption.defaultValue();
+    assertEquals(longDefaultVals.size(), 1);
+
+    conf.set(listLongConfigOption, "1,2,3");
+    List<Long> longVals = conf.get(listLongConfigOption);
+    assertEquals("[1, 2, 3]", longVals.toString());
```
Comment (jerqi) [inline common/src/test/java/com/tencent/rss/common/config/ConfigOptionTest.java:69]:
```
assertEquals(Lists.newArrayList(1L, 2L, 3L), LongVals);
```
It seems better.

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java [hunk 6]
```diff
@@ -576,7 +576,7 @@ public <T> Optional<T> getOptional(ConfigOption<T> option) {
     return value;
   }
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/RssConf.java:577]: I suggest you that you read Flink configuration implement carefully. You should also modify here.

```  
public <T> Optional<T> getOptional(ConfigOption<T> option) {
        Optional<Object> rawValue = getRawValueFromOption(option);
        Class<?> clazz = option.getClazz();

        try {
            if (option.isList()) {
                return rawValue.map(v -> ConfigurationUtils.convertToList(v, clazz));
            } else {
                return rawValue.map(v -> ConfigurationUtils.convertValue(v, clazz));
            }
        } catch (Exception e) {
            throw new IllegalArgumentException(
                    String.format(
                            "Could not parse value '%s' for key '%s'.",
                            rawValue.map(Object::toString).orElse(""), option.key()),
                    e);
        }
```

**Review by jerqi — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java (hunk 6 above)
Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java:577]: Although the RssConf impl looks like the Flink conf, but the convert method is different. That means we shouldn't do the convert action in `getOptional` method, just to init it in `TypedConfigOptionBuilder` or `ListConfigOptionBuilder`.

Now in current PR, i dont implement the case like `conf.set(listStringOption, List("a", "b", c))`. I think i can support it in the `ListConfigOptionBuilder` directly. No need to change the `getOptional` method.

Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]: And we should allow users to set conf kv in the following usages.
1. conf.set("rss.key", "a,b,c");
2. conf.set(ConfigOption<T> key, T value);

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java [hunk 7]
```diff
@@ -581,6 +581,11 @@ public <T> RssConf set(ConfigOption<T> option, T value) {
     return this;
   }
 
+  public RssConf set(String key, String value) {
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/RssConf.java:584]: We can reuse method `setString`.

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java (hunk 6 above)
Comment (jerqi) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]:
````
  public <T> Optional<T> getOptional(ConfigOption<T> option) {
    Optional<Object> rawValue = getRawValueFromOption(option);
    Class<?> clazz = option.getClazz();
    Optional<T> value = rawValue.map(v -> option.convertValue(v, clazz));
    return value;
  }
```
This is the RssConf's implement. Why do we implement the case like conf.set(listStringOption, List("a", "b", c))? 
It's not consistent with others.

**Review by jerqi — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java (hunk 7 above)
Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java:584]: Got it.

File: common/src/main/java/com/tencent/rss/common/config/RssConf.java (hunk 6 above)
Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]: This case has been supported in latest commit.

Comment (jerqi) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]: Flink Configuration's implement seems more complete. Method `StructuredOptionsSplitter` support more split characters. Should we reuse Flink's implement?

Comment (jerqi) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]: If you reuse the Flink's logic, it will also be easy to be reviewed, thanks.

Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/RssConf.java]: I think there is no need to reuse it. Just split it in ConfigOption is enough.

File: common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java [hunk 8]
```diff
@@ -54,6 +57,8 @@
  * }</pre>
  */
 public class ConfigOptions {
+  private static final String LIST_SPILTTER = ",";
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java:60]: Do we move to class ListConfigOptionBuilder?

**Review by jerqi — APPROVED** [review]
LGTM, thanks for your contribution.

File: common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java (hunk 8 above)
Comment (zuston) [reply common/src/main/java/com/tencent/rss/common/config/ConfigOptions.java]: Done

**Pull request merged**